azure-iot-device/samples/advanced-hub-scenarios/receive_direct_method.py

```diff
@@ -31,35 +31,42 @@
     # define behavior for handling methods
     async def method1_listener(device_client):
         while True:
-            method_request = await device_client.receive_method("method1")  # Wait for method1 calls
-            payload = json.dumps({"result": True, "data": "some data"})  # set response payload
+            method_request = await device_client.receive_method_request(
+                "method1"
+            )  # Wait for method1 calls
+            payload = {"result": True, "data": "some data"}  # set response payload
             status = 200  # set return status code
             print("executed method1")
-            await device_client.send_method_response(
-                method_request, payload, status
-            )  # send response
+            method_response = MethodResponse.create_from_method_request(
+                method_request, status, payload
+            )
+            await device_client.send_method_response(method_response)  # send response
```

Comment:
seeing this, i think having to have 2 steps there is a little jarring (i apologize if it was my idea and i'm changing my mind...). 
i'd rather send_method_response take just request_id, status and payload. 

i understand having a send_response method on an object called request might be weird too (and might create its own set of weird dependency/garbage collection problems)

maybe a source of inspiration would be to see how http servers serve http request/responses?

Reply:
There at least needs to be a Method Response object under the covers so the data can be wrapped up together to be used by lower layers of the stack, but I'm open to changing the API. I'll take a look at how some other libraries work, and we can revisit this before the merge.

I'm fairly against having "smart objects" at the moment, given the significant complexity that results from implementing them. But if we do decide to eventually go down that route, it would have to be an across the board change, we can't just do it for Methods, it has to be done for Messages, and any other objects in the future (Twin?).

---
In reply to: [280651175](https://github.com/Azure/azure-iot-sdk-python-preview/pull/86#discussion_r280651175) [](ancestors = 280651175)

Reply:
Follow up: I've been looking into alternatives in other libraries, but the issue is that how we handle request response is a bit weird due to how the IoTHub implementation of methods is - it doesn't map particularly well to more traditional request/response paradigms, especially since a lot of those are built around fundamental concepts of HTTP.

I think we should punt this down the road a bit. Everyone had signed off on this API over the various other samples that were presented as options, and if we want to revisit the design, I think it's a decision I'd like everyone to be a part of.

---
In reply to: [281713995](https://github.com/Azure/azure-iot-sdk-python-preview/pull/86#discussion_r281713995) [](ancestors = 281713995,280651175)